File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // brute kya hota 3 sum hai

        // set<vector<int>> triplets ;
        // int n = nums.size() ;

        // for(int i = 0 ; i < n - 2 ; i++){
        //     for(int j = i+1 ; j < n - 1 ;j++){
        //         for(int k = j+1 ; k < n ; k++){
        //             if(nums[i]+ nums[j]+nums[k] == 0){
        //                 vector<int> triplet = {nums[i],nums[j],nums[k]};
        //                 sort(triplet.begin() , triplet.end()) ;
        //                 triplets.insert(triplet);
        //             }
        //         }
        //     }
        // }
        // return vector<vector<int>>(triplets.begin(),triplets.end());

        sort(nums.begin(), nums.end());
        int n = nums.size();
        set<vector<int>> ans;

        for (int i = 0; i < nums.size() - 2; i++) {

            int j = i + 1;
            int k = n - 1;

            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];

                if (sum > 0) {
                    k--;
                } else if (sum < 0) {
                    j++;
                } else {
                    vector<int> triplet = {nums[i], nums[j], nums[k]};
                    ans.insert(triplet);
                    j++;
                    k--;
                }
            }

        }

        return vector<vector<int>>(ans.begin(),ans.end());
    }
};

File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] arr) {
        Arrays.sort(arr);
        List<List<Integer>> set = new ArrayList<>();

        for (int i = 0; i < arr.length - 2; i++) {

            if (i > 0 && arr[i] == arr[i - 1]) continue;

            int j = i + 1;
            int k = arr.length - 1;

            while (j < k) {

                int sum = arr[i] + arr[j] + arr[k];

                if (sum == 0) {

                    ArrayList<Integer> list = new ArrayList<>();
                    list.add(arr[i]);
                    list.add(arr[j]);
                    list.add(arr[k]);

                    // System.out.println(list);

                    j++;
                    k--;

                    while(j < k && arr[j] == arr[j-1])j++;
                    while(j < k && arr[k] == arr[k+1])k--;

                    set.add(list);

                } else if (sum > 0) {
                    k--;
                } else {
                    j++;
                }

            }

        }
        // System.out.println(set);
        return set ; 
    }
}